### PR TITLE
Trezor streaming

### PR DESCRIPTION
**Pull request by DavidTranDucVL**
Related PRs:
- Trezor connect: https://github.com/trezor/connect/pull/841
- Trezor firmware: https://github.com/trezor/trezor-firmware/pull/1683
  - When building trezor firmware emulator, you need to change version from 2.4.1 to 2.4.2

Tested with:
- Trezor
	- Trezor Firmware
		- https://github.com/trezor/trezor-firmware on `master` branch, at the time commit @ 6f1005b155a7d79482dfee33506f7892b4e81c48
		- Release 2.4.1: https://github.com/trezor/trezor-firmware/releases/tag/core%2Fv2.4.1
			- Transactions with regular (non-voting-registration) metadata are not backwards compatible
			- CAtalyst voting registration is failing due to zero ttl restriction, this is something to be fixed
	- Trezor Connect:
		- Prerelease 8.1.30: https://github.com/vacuumlabs/connect/releases/tag/8.1.30-rc.1-tx-streaming-extended
- Ledger
	- Cardano app Version 2.4.1 from Ledger Live
- Cardano Node
	- Release 1.29.0: https://github.com/input-output-hk/cardano-node/releases/tag/1.29.0

### CODE REVIEW

File: src/crypto-providers/trezorCryptoProvider.ts [hunk 1]
```diff
@@ -509,24 +584,32 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
     const dummyTx = prepareDummyTx(network, trezorAuxData)
 
     const response = await TrezorConnect.cardanoSignTransaction(dummyTx)
-
     if (!response.success) {
       throw Error(response.payload.error)
     }
+    if (!response?.payload?.auxiliaryDataSupplement) throw Error(Errors.MissingAuxiliaryDataSupplement)
+    if (!response?.payload?.auxiliaryDataSupplement.catalystSignature) {
```
[the remark this answers is not in the document]
Comment (DavidTranDucVL) [reply src/crypto-providers/trezorCryptoProvider.ts:592]: In the Ledger counterpart the field is not nullable. I moved part of the code to common logic.

File: src/crypto-providers/trezorCryptoProvider.ts [hunk 2]
```diff
@@ -151,6 +157,18 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
     return payload.map((result) => result.publicKey as XPubKeyHex)
   }
 
+  const determineSigningMode = (certificates: _Certificate[], signingFiles: HwSigningData[]) => {
```
[the remark this answers is not in the document]
Comment (DavidTranDucVL) [reply src/crypto-providers/trezorCryptoProvider.ts:159]: I also thought about this, but in the end I feel like it makes things more complicated without adding much value.

File: src/crypto-providers/trezorCryptoProvider.ts [hunk 3]
```diff
@@ -336,9 +354,9 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
     validityIntervalStart != null ? validityIntervalStart.toString() : undefined
   )
 
-  const prepareMetaDataHex = (metaData: Buffer | null): TrezorTypes.CardanoAuxiliaryData | undefined => (
-    metaData ? ({
-      blob: encodeCbor(metaData).toString('hex'),
+  const prepareMetaDataHex = (metaDataHash: Buffer | null): TrezorTypes.CardanoAuxiliaryData | undefined => (
```
[the remark this answers is not in the document]
Comment (DavidTranDucVL) [reply src/crypto-providers/trezorCryptoProvider.ts]: Fixed.

File: src/crypto-providers/trezorCryptoProvider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gabrielKerekes) [reply src/crypto-providers/trezorCryptoProvider.ts:159]: Ok. We'll see how this'll change with multisig and we can refactor then if needed.

**Review by gabrielKerekes — APPROVED** [review]

File: src/crypto-providers/ledgerCryptoProvider.ts [hunk 4]
```diff
@@ -646,25 +631,28 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
...
-    // TODO should be properly checked is the witness path is valid?
-    // TODO pool cold key witnesses?
-    const shelleyWitnesses = ledgerWitnesses
-      .filter((witness) => !isByronPath(witness.path))
-      .map((witness) => {
-        const { cborXPubKeyHex } = getSigningFileDataByPath(witness.path)
-        const { pubKey } = splitXPubKeyCborHex(cborXPubKeyHex)
-        return TxShelleyWitness(pubKey, witness.signature)
-      })
+    const [byronWitnesses, shelleyWitnesses] = partition(
```
Comment (janmazak) [inline src/crypto-providers/ledgerCryptoProvider.ts:634]: I think this deserves a comment on what is meant by "shelleyWitnesses". E.g. whether it includes pool cold key witnesses etc.

File: src/crypto-providers/trezorCryptoProvider.ts [hunk 5]
```diff
@@ -362,17 +381,55 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
...
+      )
+      if (hwSigningData) return hwSigningData
+      throw Error(Errors.MissingHwSigningDataAtXPubKeyError)
+    }
+
+    const [byronWitnesses, shelleyWitnesses] = partition(
+      trezorWitnesses.map((witness) => {
+        const signingFile = getSigningFileDataByXPubKey(witness.pubKey as PubKeyHex)
+        return ({
+          ...witness,
```
Comment (janmazak) [inline src/crypto-providers/trezorCryptoProvider.ts:400]: I would prefer if we were explicit about what goes into the return value. The type `TrezorTypes.CardanoSignedTxWitness` is not even part of this repository (and my VS code fails to navigate it).

The same for Ledger, I guess.

File: src/crypto-providers/trezorCryptoProvider.ts [hunk 6]
```diff
@@ -414,7 +472,36 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
...
+    const { byronWitnesses, shelleyWitnesses } = await createWitnesses(ledgerWitnesses, signingFiles)
+    const _byronWitnesses = byronWitnesses.map((byronWitness) => (
+      { key: TxWitnessKeys.BYRON, data: byronWitness }
+    ) as _ByronWitness)
+    const _shelleyWitnesses = shelleyWitnesses.map((shelleyWitness) => (
+      { key: TxWitnessKeys.SHELLEY, data: shelleyWitness }
+    ) as _ShelleyWitness)
+
+    return [..._shelleyWitnesses, ..._byronWitnesses]
   }
```
Comment (janmazak) [inline src/crypto-providers/trezorCryptoProvider.ts:505]: These functions are identical with Ledger. Don't we want to somehow unify them? (I don't see an obvious place for that right now.)

**Review by janmazak — APPROVED** [review]
I've not tried to run it or anything, but it looks good. (Maybe unifying the functions now common for ledger and trezor can be done in a separate PR if we decide to do it.)

**Pull request merged**